pandas/core/strings.py

```diff
@@ -109,9 +118,51 @@ def cat_safe(list_of_columns: List, sep: str):
 
 def _na_map(f, arr, na_result=np.nan, dtype=object):
     # should really _check_ for NA
+    if is_extension_array_dtype(arr.dtype):
```

Comment:
can you sync the signatures up with _map, also rename _map -> _map_arraylike (or similar)

Reply:
Sorry, don't follow this. I think they do match, aside from @jbrockmendel's request to call it `func` rather than `f`.

I guess `na_map` calls it `na_result` while `_map` calls it `na_value`.

Reply:
I am really -1 on 2 different branches here. If they have *exactly* the same signature a little less negative. again I would rename _map to be more informative here

Reply:
What do you mean with "-1 on 2 different branches here" ? 
The whole purpose of this PR is to add a separate branch in case of StringArray (because we can be more efficient and want to be more specific in the result dtype)

Reply:
That's why my initial PR was so much more complex, since it tried to handle both cases similarly. I think that was more complex than this.

As Joris says, the main point of divergence is that for StringArray we usually know the result dtype exactly. It doesn't depend on the presence of NAs. Additionally,

1. We're still using `map_infer_dtype` for both, so the core implementation is the same.
2. We'll eventually deprecate `.str` on object-dtype, so we will end up with just this implementation.

Reply:
ok, at the very least these signatures for what you call _map and _stringarray_map should be *exactly* the same.
and _map -> _map_object and _stringarray_map -> _map_stringarray.

I think this is crucial for not adding technical debt.

Reply:
@jreback is your suggestion to add `na_mask` to the _stringarray_map signature and have it just not be used?  I think this relates to the "this should be a StringArray method" discussion

Reply:
`_map` and `_na_map` already have inconsistent signatures. I'm not sure why it's that way on master, but I'm a bit against adding unused arguments in this case.

What's the technical debt we're adding here? By definition, we need to handle StringArray differently, since its result type doesn't depend on the presence of NAs.

Reply:
> _map and _na_map already have inconsistent signatures. I'm not sure why it's that way on master, but I'm a bit against adding unused arguments in this case.

And there is also a good reason for that, as `_map` has an additional argument `na_mask` that is used internally in `_map` (for a recursive call). 
I think refactoring `_map` is outside of the scope of this PR.

Reply:
can you create an issue to clean this up (_map and _na_map), and/or refactor of this, post this PR.